nemo/lightning/pytorch/callbacks/model_checkpoint.py

```diff
@@ -395,6 +408,25 @@ def _monitor_candidates(self, trainer: "pl.Trainer") -> Dict[str, torch.Tensor]:
 
         return monitor_candidates
 
+    def _link_checkpoint(self, trainer: "pl.Trainer", filepath: str, linkpath: str, override_async=False) -> None:
+
+        ## check to see whether this step has already been saved as top_k
+        ## in which case we can create a symlink
+        ## otherwise, we have to save the checkpoint
+        saved_current_step = str(ckpt_to_dir(linkpath)).replace("-last", "") == str(ckpt_to_dir(filepath))
+        if not saved_current_step:
+            self._save_checkpoint(trainer, linkpath)
+            return
+
+        ## linking will happen as part of the finalize fn
+        if self.async_save and not override_async:
+            self.ckpts_to_link[str(filepath)] = str(linkpath)
+            return
+
+        filepath = ckpt_to_dir(filepath)
+        linkpath = ckpt_to_dir(linkpath)
+        super()._link_checkpoint(trainer, filepath, linkpath)
+
```

Comment:
Is there a way to avoid overriding PTL's `_link_checkpoint` method ? We want to avoid overriding PTL's private methods to have stable code.

Reply:
I can think about it, but it might be challenging to support linking with async checkpointing without overwriting this method. Also, the addition of `saved_current_step` is needed to fix a bug that seems to exist in PTL's link implementation in which the `-last` checkpoint gets linked to the most recently saved checkpoint, even if that corresponds to a different step.

Reply:
I see, then maybe we should file an issue with [PTL issues](https://github.com/Lightning-AI/pytorch-lightning/issues) and ask them to fix this bug ? That way it can save us from overriding private methods.

Reply:
Yeah it would be great if PTL could fix that issue! But we'd still have to figure out how to handle async checkpointing, and I do think that would require us to overwrite either `_link_checkpoint` or `_save_last_checkpoint` (where `_link_checkpoint` is invoked)

Reply:
Agree that for handling async save we have to override `_link_checkpoint` anyway.
But since we call `super()._link_checkpoint` I don't think there is too much risk connected with that